### PR TITLE
feat(app): introduce single instance app

### PR DESCRIPTION
**Pull request by Phil91**
add endpoint to add app instance information
edit activate app process
edit offer subscription process for single instance apps

Refs: CPLP-1846

### CODE REVIEW

File: src/portalbackend/PortalBackend.PortalEntities/Entities/AppInstanceSetup.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    public bool IsSingleInstance { get; set; }
+    
+    public string? InstanceUrl { get; set; }
+
+    public string? ServiceAccountId { get; set; }
+    
+    // Navigation properties
+    public virtual Offer? App { get; private set; }
+
+    public virtual IamServiceAccount? ServiceAccount { get; private set; }
```
Comment (ntruchsess) [inline src/portalbackend/PortalBackend.PortalEntities/Entities/AppInstanceSetup.cs:49]: From my point of view this should rather link the companyServiceAccount not the iamServiceAccount (iamService-account is designed being a pure link in between companyServiceAccount and keycloak)
I'm also not sure whether the AppInstanceSetup entity should link the technical user at all. Isn't appInstanceSetup a container that defines this app should be set up as singleinstance app so it contains all data that is required to do so? From my point of view the technical user is the result of this setup and should be linked to the actual app-instance not the setup-data.

**Review by ntruchsess — CHANGES_REQUESTED** [review]
I did some minor cleanup and rebased on dev resolving the conflicts that resulted from the merge or Majids PR. Please make sure to get those changes before proceeding

File: src/portalbackend/PortalBackend.DBAccess/Repositories/OfferRepository.cs [hunk 2]
```diff
@@ -670,4 +676,55 @@ sorting switch
...
+        var appInstanceSetup = _context.Attach(entity).Entity;
+        setOptionalParameters.Invoke(appInstanceSetup);
+    }
+
+    /// <inheritdoc />
+    public Task<SingleInstanceOfferData?> GetSingleInstanceOfferData(Guid offerId, OfferTypeId offerTypeId) =>
+        _context.Offers.Where(o => o.Id == offerId && o.OfferTypeId == offerTypeId)
+            .Select(o => new
+            {
+                Company = o.ProviderCompany, OfferId = o.Id, OfferName = o.Name,
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/PortalBackend.DBAccess/Repositories/OfferRepository.cs]: absolutely right, removed it.

File: src/portalbackend/PortalBackend.PortalEntities/Entities/IamClient.cs [hunk 3]
```diff
@@ -42,5 +42,7 @@ public IamClient(Guid id, string clientClientId) : this()
     [StringLength(255)]
     public string ClientClientId { get; private set; }
 
+    public bool Disabled { get; set; }
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/portalbackend/PortalBackend.PortalEntities/Entities/IamClient.cs]: removed from the database

File: src/portalbackend/PortalBackend.PortalEntities/Entities/AppInstanceSetup.cs (hunk 1 above)
Comment (Phil91) [reply src/portalbackend/PortalBackend.PortalEntities/Entities/AppInstanceSetup.cs]: changed it to companyServiceAccount and moved it to the `AppInstance`

**Comment by Phil91** [comment]
@ntruchsess did the change regarding the technical user profile. it's ready to be reviewed again

File: src/marketplace/Offers.Library/Service/OfferSetupService.cs [hunk 4]
```diff
@@ -125,6 +134,83 @@ public async Task<OfferAutoSetupResponseData> AutoSetupOfferAsync(OfferAutoSetup
...
+                data.Bpn == null ? Enumerable.Empty<string>() : Enumerable.Repeat(data.Bpn, 1),
+                CompanyServiceAccountTypeId.MANAGED,
+                data.EnhanceTechnicalUserName,
+                sa => { sa.OfferSubscriptionId = subscriptionId; })
+            .ConfigureAwait(false);
+        return new TechnicalUserInfoData(serviceAccountId, serviceAccountData.AuthData.Secret, technicalClientId);
+    }
+
+    /// <inheritdoc />
+    public async Task SetupSingleInstance(Guid offerId, string instanceUrl)
```
Comment (ntruchsess) [inline src/marketplace/Offers.Library/Service/OfferSetupService.cs:163]: that method should validate there is no preexisting appinstance before creating a new one.

**Review by ntruchsess — CHANGES_REQUESTED** [review]
I did some cleanup rebasing on latest dev. Please make sure to get those updates before proceeding.

File: src/marketplace/Offers.Library/Service/OfferSetupService.cs (hunk 4 above)
Comment (Phil91) [reply src/marketplace/Offers.Library/Service/OfferSetupService.cs:163]: valid point 👍  added

File: src/marketplace/Offers.Library/Service/OfferSetupService.cs [hunk 5]
```diff
@@ -125,6 +134,83 @@ public async Task<OfferAutoSetupResponseData> AutoSetupOfferAsync(OfferAutoSetup
...
+    {
+        var userRolesRepository = _portalRepositories.GetInstance<IUserRolesRepository>();
+        var (_, iamClientId) = await CreateClient(instanceUrl, offerId, false, userRolesRepository);
+        _portalRepositories.GetInstance<IAppInstanceRepository>().CreateAppInstance(offerId, iamClientId);
+    }
+
+    /// <inheritdoc />
+    public async Task DeleteSingleInstance(Guid appInstanceId, Guid clientId, string clientClientId)
+    {
+        await _provisioningManager.DeleteCentralClientAsync(clientClientId)
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/marketplace/Offers.Library/Service/OfferSetupService.cs:180]: At this point there shouldn't be any technical users assigned to the app instance (at least process wise that's not possible). If added the delete for it anyway.
In my point of view, we should somehow add a generic logic for our keycloak calls where we could do rollbacks if failing

**Review by ntruchsess — CHANGES_REQUESTED** [review]

File: src/marketplace/Offers.Library/Service/OfferSetupService.cs [hunk 6]
```diff
@@ -125,6 +134,101 @@ public async Task<OfferAutoSetupResponseData> AutoSetupOfferAsync(OfferAutoSetup
...
+        }
+
+        var (technicalClientId, serviceAccountData, serviceAccountId, _) = await _serviceAccountCreation
+            .CreateServiceAccountAsync(
+                technicalUserInfoCreations.Single(),
+                data.CompanyId,
+                data.Bpn == null ? Enumerable.Empty<string>() : Enumerable.Repeat(data.Bpn, 1),
+                CompanyServiceAccountTypeId.MANAGED,
+                data.EnhanceTechnicalUserName,
+                sa => { sa.OfferSubscriptionId = subscriptionId; })
```
[the remark this answers is not in the document]
Comment (Phil91) [reply src/marketplace/Offers.Library/Service/OfferSetupService.cs:157]: added the coverage for that

**Comment by ntruchsess** [comment]
lgtm. Will approve as soon it is rebased and has the Migration recreated after merge of https://github.com/catenax-ng/tx-portal-backend/pull/301

**Comment by Phil91** [comment]
> lgtm. Will approve as soon it is - after merge of #301 - rebased and has the Migration recreated

will do the migration recreation 👍

**Review by ntruchsess — APPROVED** [review]
did rebase and recreate the migration

**Pull request merged**